app/scripts/metamask-controller.js

```diff
@@ -738,10 +774,76 @@ export default class MetamaskController extends EventEmitter {
       );
     });
 
+    this.setupControllerEventSubscriptions();
+
     // TODO:LegacyProvider: Delete
     this.publicConfigStore = this.createPublicConfigStore();
   }
 
+  /**
+   * Sets up BaseController V2 event subscriptions. Currently, this includes
+   * the subscriptions necessary to notify permission subjects of account
+   * changes.
+   *
+   * Some of the subscriptions in this method are ControllerMessenger selector
+   * event subscriptions. See the relevant @metamask/controllers documentation
+   * for more information.
+   *
+   * Note that account-related notifications emitted when the extension
+   * becomes unlocked are handled in MetaMaskController._onUnlock.
+   */
+  setupControllerEventSubscriptions() {
+    const getCurrentAccounts = async (origin, newAccounts) => {
+      return newAccounts.length < 2
+        ? newAccounts
+        : await this.getPermittedAccounts(origin);
+    };
+
+    const handleAccountsChange = async (origin, newAccounts) => {
+      this.notifyConnections(origin, {
+        method: NOTIFICATION_NAMES.accountsChanged,
+        params: await getCurrentAccounts(origin, newAccounts),
+      });
+      this.permissionLogController.updateAccountsHistory(origin, newAccounts);
+    };
+
+    // This handles account changes whenever the selected address changes.
+    let lastSelectedAddress;
+    this.preferencesController.store.subscribe(async ({ selectedAddress }) => {
+      if (selectedAddress && selectedAddress !== lastSelectedAddress) {
+        lastSelectedAddress = selectedAddress;
+        const permittedAccountsMap = getPermittedAccountsByOrigin(
+          this.permissionController.state,
+        );
+
+        for (const [origin, accounts] of permittedAccountsMap.entries()) {
+          if (accounts.includes(selectedAddress)) {
+            handleAccountsChange(origin, accounts);
+          }
+        }
+      }
+    });
+
+    // This handles account changes every time relevant permission state
+    // changes, for any reason.
+    this.controllerMessenger.subscribe(
+      `${this.permissionController.name}:stateChange`,
+      async (currentValue, previousValue) => {
+        if (this.isUnlocked()) {
```

Comment:
Presumably we still want the `this.permissionLogController.updateAccountsHistory(origin, newAccounts);` call to happen even if the wallet is locked. That call is part of `handleAccountsChange`.

Reply:
Fixed here: 83502cbf8